charms/jupyter-ui/src/charm.py

```diff
@@ -25,139 +39,191 @@ def __init__(self, msg, status_type=None):
         self.status = status_type(msg)
 
 
-class Operator(CharmBase):
-    """Charm for the Jupyter UI."""
+class JupyterUI(CharmBase):
+    """A Juju Charm for Jupyter UI."""
 
     def __init__(self, *args):
-        """Charm for the Jupyter UI."""
+        """Initialize charm and setup the container."""
         super().__init__(*args)
 
-        self.log = logging.getLogger(__name__)
+        # retrieve configuration and base settings
+        self.logger = logging.getLogger(__name__)
+        self._namespace = self.model.name
+        self._lightkube_field_manager = "lightkube"
+        self._name = self.model.app.name
+        self._http_port = self.model.config["port"]
+        self._exec_command = (
+            "gunicorn"
+            " -w 3"
+            f" --bind 0.0.0.0:{self._http_port}"
+            " --access-logfile"
+            " - entrypoint:app"
+        )
+        self._container_name = "jupyter-ui"
+        self._container = self.unit.get_container(self._name)
+
+        # setup context to be used for updating K8S resources
+        self._context = {
+            "app_name": self._name,
+            "namespace": self._namespace,
+            "service": self._name,
+        }
+        self._k8s_resource_handler = None
 
-        self.image = OCIImageResource(self, "oci-image")
+        http_port = ServicePort(int(self._http_port), name="http")
+        self.service_patcher = KubernetesServicePatch(
+            self, [http_port], service_name=f"{self.model.app.name}"
+        )
+
+        # setup events
         for event in [
-            self.on.install,
             self.on.leader_elected,
             self.on.upgrade_charm,
             self.on.config_changed,
             self.on["ingress"].relation_changed,
+            self.on.jupyter_ui_pebble_ready,
         ]:
             self.framework.observe(event, self.main)
+        self.framework.observe(self.on.jupyter_ui_pebble_ready, self._on_pebble_ready)
```

Comment:
Given that we are calling `main` in `_on_pebble_ready`, why are we adding `self.on.jupyter_ui_pebble_ready` in the for loop also?